mingw-w64-git/git-wrapper.c

```diff
@@ -802,9 +802,16 @@ int main(void)
 		if (!wcsicmp(prefix_args + prefix_args_len - 4, L".exe"))
 			prefix_args_len -= 4;
 
+		initialize_top_level_path(top_level_path, exepath, msystem_bin, 1);
+
 		/* set the default exe module */
-		wcscpy(exe, exepath);
+		wcscpy(exe, top_level_path);
+		my_path_append(exe, msystem_bin, MAX_PATH);
 		my_path_append(exe, L"git.exe", MAX_PATH);
+		if (_waccess(exe, 0) == -1) {
+			wcscpy(exe, top_level_path);
+			my_path_append(exe, L"bin\\git.exe", MAX_PATH);
+		}
```

Comment:
Hmm. It is not quite clear to me what/how the problem is that this diff hunk is trying to solve (read: what situation does the existing code not handle well, and how does the new code do better)?

Please do follow [this advice](https://github.blog/2022-06-30-write-better-commits-build-better-projects/) how to include valuable information in the commit message.

In this instance, I _guess_ you're trying to make it possible for `git-wrapper.exe` to be copied to `<Git>\cmd\git-receive-pack.exe` so that it hands off to the appropriate `git.exe` when being called as `git-receive-pack.exe`? If so:

- Why does it not work as expected _already_? It should hand off to `cmd\git.exe receive-pack` which in turn should hand off to `mingw64\bin\git.exe` and all is fine and dandy.
- Even if it is not fine and dandy, the code path that is modified here supports the case where the Git wrapper is copied into `mingw64\libexec\git-core\git-<builtin>.exe`, and that code path is modified, which makes me suspect that it is getting broken in the process.

Reply:
> Why does it not work as expected already? It should hand off to cmd\git.exe receive-pack which in turn should hand off to mingw64\bin\git.exe and all is fine and dandy.

From what I can see it seems to call `cmd\git.exe` with `argv[0]` replaced with `git-receive-pack`, but it is ignored afterwards and `git` proper is called instead.

It will certainly be better if this can be fixed.

Reply:
> Even if it is not fine and dandy, the code path that is modified here supports the case where the Git wrapper is copied into mingw64\libexec\git-core\git-<builtin>.exe, and that code path is modified, which makes me suspect that it is getting broken in the process.

I actually haven't looked into this, considering this I think trying to resolve the first issue would be better.

Reply:
> From what I can see it seems to call cmd\git.exe with argv[0] replaced with git-receive-pack, but it is ignored afterwards and git proper is called instead.

Yeah this is exactly the case.

`cmd/git-receive-pack.exe -h` gets executed => 
executes `cmd\git.exe` with `lpCommandLine` in `CreateProcess` set as `git-receive-pack -h` => 

`cmd\git.exe` gets the commandline, but drops argv[0] (`git-receive-pack`) and replace with `git.exe` =>
execute `bin\git.exe` with the commandline (`git.exe -h`) =>

git proper gets executed instead, in this case shows git help.

Looking at `fixup_commandline`, it handles argv[0] differently on whether `prefix_args` is null. In `call a builtin`, this is stripped from `basename`.

In the base `call git.exe` branch, `basename` would be `git.exe`. I guess I can get `prefix_args` from argv[0] (need to parse the commandline ourselves though), but is it safe to do this unconditionally? If not, what condition should I check?